activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb

```diff
@@ -639,11 +639,28 @@ def alter_table(
             yield definition if block_given?
           end
 
-          transaction do
-            disable_referential_integrity do
+          fk_enabled = query_value("PRAGMA foreign_keys") == 1
+
+          begin
+            execute("PRAGMA foreign_keys = OFF")
+
+            transaction do
               move_table(table_name, altered_table_name, options.merge(temporary: true))
               move_table(altered_table_name, table_name, &caller)
             end
+
+            execute("PRAGMA foreign_keys = ON") if fk_enabled
+
+            if fk_enabled
+              violations = execute("PRAGMA foreign_key_check").to_a
+              if violations.any?
+                raise ActiveRecord::StatementInvalid,
+                      "Foreign key violations detected after schema change: #{violations.inspect}"
+              end
+            end
+          rescue => _
+            execute("PRAGMA foreign_keys = ON") if fk_enabled
```

Comment:
Would not this execute `execute("PRAGMA foreign_keys = ON")` again if there is any violation?

Reply:
It won't be required anymore, @rafaelfranca :)